cpp/src/arrow/dataset/scanner.cc

```diff
@@ -66,53 +119,96 @@ ScanTaskIterator SimpleScanner::Scan() {
+                                       ScanTaskIterator* out) {
+  std::unique_ptr<ScanTask> scan_task = internal::make_unique<SimpleScanTask>(batches);
+
+  std::vector<std::unique_ptr<ScanTask>> tasks;
+  tasks.emplace_back(std::move(scan_task));
+
+  *out = MakeVectorIterator(std::move(tasks));
+  return Status::OK();
 }
 
 ScannerBuilder::ScannerBuilder(std::shared_ptr<Dataset> dataset,
                                std::shared_ptr<ScanContext> scan_context)
     : dataset_(std::move(dataset)),
       scan_options_(ScanOptions::Defaults()),
-      scan_context_(std::move(scan_context)) {}
+      scan_context_(std::move(scan_context)),
+      pool_(default_memory_pool()) {}
+
+Status EnsureColumnsInSchema(const std::shared_ptr<Schema>& schema,
+                             const std::vector<std::string>& columns) {
+  for (const auto& column : columns) {
+    if (schema->GetFieldByName(column) == nullptr) {
+      return Status::Invalid("Requested column ", column,
+                             " not found in dataset's schema.");
+    }
+  }
 
-ScannerBuilder* ScannerBuilder::Project(const std::vector<std::string>& columns) {
-  return this;
+  return Status::OK();
 }
 
-ScannerBuilder* ScannerBuilder::Filter(std::shared_ptr<Expression> filter) {
-  scan_options_->filter = std::move(filter);
-  return this;
+Status ScannerBuilder::Project(const std::vector<std::string>& columns) {
+  RETURN_NOT_OK(EnsureColumnsInSchema(schema(), columns));
+  project_columns_ = columns;
+  return Status::OK();
 }
 
-ScannerBuilder* ScannerBuilder::Filter(const Expression& filter) {
-  return Filter(filter.Copy());
+Status ScannerBuilder::Filter(std::shared_ptr<Expression> filter) {
+  if (filter != nullptr) {
+    RETURN_NOT_OK(EnsureColumnsInSchema(schema(), FieldsInExpression(*filter)));
+  }
+  scan_options_->filter = std::move(filter);
+  return Status::OK();
 }
 
-ScannerBuilder* ScannerBuilder::FilterEvaluator(
-    std::shared_ptr<ExpressionEvaluator> evaluator) {
-  scan_options_->evaluator = std::move(evaluator);
-  return this;
+Status ScannerBuilder::SetMemoryPool(MemoryPool* pool) {
+  pool_ = pool;
+  return Status::OK();
 }
 
-ScannerBuilder* ScannerBuilder::SetGlobalFileOptions(
-    std::shared_ptr<FileScanOptions> options) {
-  return this;
-}
+Status ScannerBuilder::Filter(const Expression& filter) { return Filter(filter.Copy()); }
 
-ScannerBuilder* ScannerBuilder::IncludePartitionKeys(bool include) {
-  scan_options_->include_partition_keys = include;
-  return this;
+std::shared_ptr<Schema> SchemaFromColumnNames(
+    const std::shared_ptr<Schema>& input, const std::vector<std::string>& column_names) {
+  std::vector<std::shared_ptr<Field>> columns;
+  for (const auto& name : column_names) {
+    columns.push_back(input->GetFieldByName(name));
+  }
+
+  return std::make_shared<Schema>(columns);
 }
 
 Status ScannerBuilder::Finish(std::unique_ptr<Scanner>* out) const {
+  if (!project_columns_.empty()) {
```

Comment:
If I'm reading this right, then `project_columns_.empty()` is equivalent to `SELECT *`. However one of the test cases indicates that projection of *no* columns is valid and also signalled by empty project columns. I think you need a boolean flag or so to explicitly indicate that default (`*`) projection is in effect